### ISSUE TITLE
What is the best way for add negative interactions?

### ISSUE THREAD

**Issue by MaxBalashov** [lyst/lightfm#204]
I saw in some problem discussion (https://github.com/lyst/lightfm/issues/180#issuecomment-302106090, last maciejkula's message)  that:
- 1s are positive;
- 0s are negative.

But guys from Catalant ([Using LightFM to Recommend Projects to Consultants](https://medium.com/product-at-catalant-technologies/using-lightfm-to-recommend-projects-to-consultants-44084df7321c)) use:
- 1s for positive;
- -1s for negative.

What is better and why?

Best regards,
Max Balashov

**Comment by maciejkula** [comment]
You can use implicit feedback (1s/0s) for the WARP and BPR losses, and explicit feedback (1s and -1s) for the logistic loss.

Which ones is better depends on your application. I would define an objective function that works for you, and evaluate both. For example, you may want to balance ranking positive items highly (good) and ranking known negative items highly (bad), and select model hyperparameters based on that.

**Comment by MaxBalashov** [comment]
Unfortunately, for my case using different values for positive and negative interactions doesn't effect on auc score (and other metrics), it's constant. How should it be interpreted?

**Comment by maciejkula** [comment]
If you're model's performance is constant regardless of what you do that suggests there is something wrong with the model.

Are the predictions `NaN`s? Are they all zeros?

**Comment by MaxBalashov** [comment]
It is my embeddings and biases. It's not `NaN`s and zeros. But  `model.user_biases` looks strange.
Sorry for so large code outputs.
```

> model.user_embeddings

> array([[0.01854292, -0.03616943,  0.02743073, ..., -0.08873698,
        -0.0281186 ,  0.04932814],
       [ 0.01835506, -0.02285067, -0.02461982, ..., -0.0196003 ,
        -0.00183558, -0.01746818],
       [-0.00016604, -0.01641887,  0.00801281, ...,  0.02026852,
        -0.01372182, -0.01250093],
       ..., 
       [-0.00038331, -0.02331057,  0.0014284 , ..., -0.0269407 ,
        -0.02862924,  0.04169686],
       [ 0.13069977,  0.01168023, -0.0036469 , ..., -0.09148387,
        -0.0869351 , -0.00417987],
       [-0.00157261,  0.01230439,  0.00775047, ..., -0.02291745,
         0.00544394, -0.01943119]], dtype=float32)

> model.item_embeddings

> array([[0.02933569, -0.0732343 ,  0.07156094, ..., -0.09587108,
        -0.02510785,  0.04858794],
       [-0.02053796, -0.01279098,  0.00457765, ...,  0.01316937,
         0.0132997 ,  0.01030643],
       [ 0.0673373 ,  0.01364722, -0.00136843, ..., -0.00953384,
        -0.00528875, -0.01918415],
       ..., 
       [ 0.15962589, -0.32825601,  0.30606297, ..., -0.20719305,
        -0.12428237,  0.2490337 ],
       [-0.0179254 ,  0.04813654, -0.07512292, ...,  0.1126662 ,
         0.11551356, -0.07269695],
       [-0.316017  ,  0.2270131 , -0.19657041, ...,  0.46460572,
         0.20413968, -0.25520191]], dtype=float32)

> model.user_biases

> array([-0.07031802,  0.        ,  0.        ,  0.        , -0.04835589,
        0.        ,  0.        ,  0.        ,  0.        ,  0.        ,
        0.        ,  0.        , -0.73506993,  0.        , -0.05971936,
        0.        , -0.00876252, -0.00447214, -0.0531995 , -0.06756987,
        0.        ,  0.        ,  0.        ,  0.        ,  0.        ,
        0.        ,  0.        ,  0.        ,  0.        ,  0.        ,
        ...,
        0.        ,  0.        ,  0.        ,  0.        ,  0.        ,
        0.        ,  0.        ,  0.        ,  0.        ,  0.        ,
       -0.00447214,  0.        , -0.01356889,  0.        ,  0.        ,
        0.        ,  0.        , -0.02967845,  0.        ,  0.        ,
        0.        ,  0.        ,  0.        ,  0.        , -0.026685  ,
       -0.15701669,  0.        ], dtype=float32)

> model.item_biases

> array([3.45653705e-02,   8.24284647e-03,   1.79438330e-02,
        -2.43308563e-02,   1.47444410e-02,   1.29504711e-03,
        -1.56834326e-03,   1.76689308e-02,  -3.99955250e-02,
         1.15162646e-02,  -3.82953659e-02,  -8.33974569e-04,
         3.71466838e-02,  -1.56373773e-02,   2.79283850e-03,
         2.29972787e-02,   0.00000000e+00,   5.43882977e-03,
         ...,
        -5.75311063e-03,  -1.46178296e-02,   5.30885439e-03,
         2.16178242e-02,  -6.85118930e-03,  -8.92395154e-03,
        -5.27057843e-03,   2.16984842e-02,   4.13220096e-03,
        -1.97833162e-02,   3.11827380e-02,  -2.06727460e-02,
         1.01115098e-02,  -1.69814434e-02,   0.00000000e+00,
         2.73628443e-01,   8.90778452e-02,  -4.03039247e-01], dtype=float32)
```

**Comment by maciejkula** [comment]
Coordinates (that is, which user interacted with which item) are definitely key.

**Comment by MaxBalashov** [comment]
Could you give advice about scaling input data? Should I scaling user and item features for avoiding overfitting?
Usually matrix factorizations depends on it, as example SVD.

**Comment by maciejkula** [comment]
You should definitely scale your input data. Just experiment with getting the right value; between 0 and 1 is a good start.

This is not as much about overfitting as about making sure the training doesn't diverge: inputs that are large will likely lead to poor fit or outright numerical failures.
